### PR TITLE
Allow specifying stepDuration as a map.

### LINKED ISSUE #122: Changing the value of `stepDuration` option in `spec.json` for AuRa
Please figure out how we could create an option `stepDurationTransition` for Parity client which would allow us to change block time for an existing PoA chain at the specified future block (simultaneously for all validators). This is for changing the value of `stepDuration` param in `spec.json`. For example: https://github.com/poanetwork/poa-chain-spec/blob/e36f2147948b1e84541f96d37ad154725d64348c/spec.json#L6

### PR DESCRIPTION
**Pull request by afck**
Closes #122, #124, #150.

### CODE REVIEW

File: ethcore/src/engines/authority_round/mod.rs [hunk 1]
```diff
@@ -700,23 +757,25 @@ impl<'a, A: ?Sized, B> Deref for CowLike<'a, A, B> where B: AsRef<A> {
 impl AuthorityRound {
 	/// Create a new instance of AuthorityRound engine.
 	pub fn new(our_params: AuthorityRoundParams, machine: EthereumMachine) -> Result<Arc<Self>, Error> {
-		if our_params.step_duration == 0 {
-			error!(target: "engine", "Authority Round step duration can't be zero, aborting");
-			panic!("authority_round: step duration can't be zero")
+		if !our_params.step_durations.contains_key(&0) {
+			error!(target: "engine", "Authority Round step 0 duration is undefined, aborting");
+			panic!("authority_round: step 0 duration is undefined")
```
Comment (vkomenda) [inline ethcore/src/engines/authority_round/mod.rs]: "step 0" -> "timestamp 0"

File: ethcore/src/engines/authority_round/mod.rs [hunk 2]
```diff
@@ -61,12 +62,13 @@ pub type RandomnessPhaseError = randomness::PhaseError;
 
 /// `AuthorityRound` params.
 pub struct AuthorityRoundParams {
-	/// Time to wait before next block or authority switching,
-	/// in seconds.
+	/// A map defining intervals of blocks with the given times (in seconds) to wait before next
+	/// block or authority switching. The keys in the map are numbers of starting blocks of those
```
Comment (vkomenda) [inline ethcore/src/engines/authority_round/mod.rs]: "numbers of starting blocks" -> "timestamps of starting blocks"

**Comment by varasev** [comment]
Should the timestamps be calculated in UTC or in the local time zone?

File: ethcore/src/engines/authority_round/mod.rs (hunk 1 above)
Comment (afck) [reply ethcore/src/engines/authority_round/mod.rs]: Not sure; timestamp 0 is always also step 0, so this is the duration of step 0.

File: ethcore/src/engines/authority_round/mod.rs (hunk 2 above)
Comment (afck) [reply ethcore/src/engines/authority_round/mod.rs]: Done. :+1:

**Comment by afck** [comment]
UTC. The timestamps are in [Unix time](https://en.wikipedia.org/wiki/Unix_time).

File: ethcore/src/engines/authority_round/mod.rs [hunk 3]
```diff
@@ -61,12 +62,13 @@ pub type RandomnessPhaseError = randomness::PhaseError;
...
 pub struct AuthorityRoundParams {
-	/// Time to wait before next block or authority switching,
-	/// in seconds.
+	/// A map defining intervals of blocks with the given times (in seconds) to wait before next
+	/// block or authority switching. The keys in the map are timestamps of starting blocks of those
+	/// periods. The entry at `0` should be defined.
 	///
-	/// Deliberately typed as u16 as too high of a value leads
-	/// to slow block issuance.
-	pub step_duration: u16,
```
Comment (varasev) [inline ethcore/src/engines/authority_round/mod.rs]: We need to keep the `step_duration` parameter as well for backward compatibility.

**Comment by varasev** [comment]
If I remove the redundant `current_duration`, `starting_sec`, `starting_step` fields from the `Step` structure, the unit tests `test_counter_increment_too_high` and `test_counter_duration_remaining_too_high` in `ethcore/src/engines/authority_round/mod.rs` are fine.

File: ethcore/src/engines/authority_round/mod.rs (hunk 3 above)
Comment (afck) [reply ethcore/src/engines/authority_round/mod.rs]: That should still be accepted: It can now either be a number, like before, or a map.

**Comment by afck** [comment]
Sorry! I updated the unit tests.

File: ethcore/src/engines/authority_round/mod.rs (hunk 3 above)
Comment (varasev) [reply ethcore/src/engines/authority_round/mod.rs]: Ah, I see, OK.

**Pull request merged**